### DISCUSSION TITLE
New musical font system for text

### DISCUSSION THREAD

**Discussion by craigsapp** [rism-digital/verovio#3052] [Q&A] [answered]
Here is some sample music:

<img width="185" alt="Screen Shot 2022-10-04 at 5 02 37 PM" src="https://user-images.githubusercontent.com/3487289/193952307-645fbfd9-415a-4243-b945-3496c6186942.png">

<details>
<summary> Click to view MEI data for above example </summary>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="5.0.0-dev">
 <meiHead>
  <fileDesc>
   <titleStmt>
    <title />
   </titleStmt>
   <pubStmt />
  </fileDesc>
  <encodingDesc>
   <appInfo>
    <application isodate="2022-10-04T16:55:09" version="3.11.0-dev-a5b9b01-dirty">
     <name>Verovio</name>
     <p>Transcoded from Humdrum</p>
    </application>
   </appInfo>
  </encodingDesc>
  <workList>
   <work>
    <title />
   </work>
  </workList>
 </meiHead>
 <music>
  <body>
   <mdiv xml:id="mz1c4a1">
    <score xml:id="s1us91oz">
     <scoreDef xml:id="s6syghj">
      <staffGrp xml:id="s1189ijp">
       <staffDef xml:id="staffdef-L2F1" n="1" lines="5">
        <clef xml:id="c2ir7lc" shape="G" line="2" />
       </staffDef>
      </staffGrp>
     </scoreDef>
     <section xml:id="section-L2F1">
      <measure xml:id="measure-L1" n="1">
       <staff xml:id="staff-L2F1" n="1">
        <layer xml:id="layer-L1F1N1" n="1">
         <note xml:id="note-L4F1" dur="1" oct="4" pname="c" accid.ges="n" />
        </layer>
       </staff>
       <tempo xml:id="tempo-L1F1" staff="1" tstamp="1.000000">
        <rend xml:id="r19c44uo" fontname="VerovioText"></rend> = 152</tempo>
      </measure>
     </section>
    </score>
   </mdiv>
  </body>
 </music>
</mei>
```

</details>

Here is the `<tempo>` element:

```xml
       <tempo xml:id="tempo-L1F1" staff="1" tstamp="1.000000">
        <rend xml:id="r19c44uo" fontname="VerovioText"></rend> = 152</tempo>
```

where the unprintable character is the bytes  ee 87 95.

The above MEI data generates an SVG with no errors when using verovio from a few months ago (see graphical example above).  When running this MEI data through the most recent develop version of verovio, I get this warning and error:

```
[Warning] Using rend@fontname with 'VerovioText' is deprecated. Use 'rend@fontfam="smufl"' instead
Segmentation fault: 11
```

Here is the stack at the point of the segmentation fault:

```
_platform_strlen + 18
__vfprintf + 5379
__v2printf + 475
_vsnprintf + 417
vsnprintf + 68
vrv::LogWarning(char const*, ...) + 253 
vrv::SvgDeviceContext::Commit(bool) + 1093
vrv::SvgDeviceContext::GetStringSVG(bool) + 33
vrv::Toolkit::RenderToSVG(int, bool) + 603
vrv::Toolkit::RenderToSVGFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) + 98
main + 6610                             
start + 1
```

Converting the same Humdrum data:

```tsv
!!!OMD: [quarter]=152
**kern
=1-
1c
=
*-
```

with the latest develop version of verovio produces the following slightly different MEI data:

<details>
<summary> Click to view newer MEI conversion </summary>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="5.0.0-dev">
   <meiHead>
      <fileDesc>
         <titleStmt>
            <title />
         </titleStmt>
         <pubStmt />
      </fileDesc>
      <encodingDesc>
         <appInfo>
            <application isodate="2022-10-04T16:53:43" version="3.13.0-dev-d45c286">
               <name>Verovio</name>
               <p>Transcoded from Humdrum</p>
            </application>
         </appInfo>
      </encodingDesc>
      <workList>
         <work>
            <title />
         </work>
      </workList>
      <extMeta>
         <frames xmlns="http://www.humdrum.org/ns/humxml">
            <metaFrame n="0" token="!!!OMD: [quarter]=152" xml:id="L1">
               <frameInfo>
                  <startTime float="0" />
                  <frameType>reference</frameType>
                  <referenceKey>OMD</referenceKey>
                  <referenceValue>[quarter]=152</referenceValue>
               </frameInfo>
            </metaFrame>
         </frames>
      </extMeta>
   </meiHead>
   <music>
      <body>
         <mdiv xml:id="marrfqb">
            <score xml:id="sgv0ak6">
               <scoreDef xml:id="sl5qw06" midi.bpm="400.000000">
                  <staffGrp xml:id="sn5w02e">
                     <staffDef xml:id="staffdef-L2F1" n="1" lines="5">
                        <clef xml:id="cdeby5d" shape="G" line="2" />
                     </staffDef>
                  </staffGrp>
               </scoreDef>
               <section xml:id="section-L2F1">
                  <measure xml:id="measure-L1" n="1">
                     <staff xml:id="staff-L2F1" n="1">
                        <layer xml:id="layer-L1F1N1" n="1">
                           <note xml:id="note-L4F1" dur="1" oct="4" pname="c" accid.ges="n" />
                        </layer>
                     </staff>
                     <tempo xml:id="tempo-L1F1" staff="1" tstamp="1.000000">
                        <rend xml:id="rmo6bdu" fontfam="smufl"></rend> = 152</tempo>
                  </measure>
               </section>
            </score>
         </mdiv>
      </body>
   </music>
</mei>
```

</details>

Where the `<tempo>` element being different (probably with additions by @lpugin to iohumdrum.cpp this summer):

```xml
                     <tempo xml:id="tempo-L1F1" staff="1" tstamp="1.000000">
                        <rend xml:id="rmo6bdu" fontfam="smufl"></rend> = 152</tempo>
```

This unprintable character are the bytes: ee 87 95 

The warning message about `@fontfam` is now gone since the newest develop version is producing `@fontfam="smufl"`, but the segmentation fault still remains with the same stack:

```
_platform_strlen + 18
__vfprintf + 5379
__v2printf + 475
_vsnprintf + 417
vsnprintf + 68
vrv::LogWarning(char const*, ...) + 253
vrv::SvgDeviceContext::Commit(bool) + 1093
vrv::SvgDeviceContext::GetStringSVG(bool) + 33
vrv::Toolkit::RenderToSVG(int, bool) + 603
vrv::Toolkit::RenderToSVGFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) + 98
main + 6610
start + 1
```

with no SVG image produced.

--------------------------------

So there are two questions/comments:

(1) Is there a better way or more correct way of encoding quarter note equals 152 compared to the current encoding:

```xml
                     <tempo xml:id="tempo-L1F1" staff="1" tstamp="1.000000">
                        <rend xml:id="rmo6bdu" fontfam="smufl"></rend> = 152</tempo>
```

I presume the unprintable bytes for the quarter note (ee 87 95) are UTF-8.  Perhaps that is being converted incorrectly (to a new UTF-32 system)?   Or is there a more human-readable way of encoding the quarter note?

And (2) in any case, the data should not crash verovio.  The function that crashes seems to be `vrv::LogWarning`'s use of `vsnprintf`, and perhaps it is trying to print the quarter note character bytes (or their conversion to UTF-32) but is choking on them.

**Comment by lpugin** [comment]
You can use entities for encoding the quarter note. 
```xml
<tempo xml:id="tempo-L1F1" staff="1" tstamp="1.000000">
   <rend fontfam="smufl">&#xECA5;</rend> = 152</tempo>
```
If you are outputting MEI from Verovio, you need to set the `--output-smufl-xml-entities` option for them to be preserved.


The other possibility (version pushed now) is to use `<symbol>`, which means that you can set the code in `glyph.num`:
```xml
<tempo xml:id="tempo-L1F1" staff="1" tstamp="1.000000">
   <symbol glyph.num="U+ECA5" glyph.auth="smufl"/> = 152</tempo>
```
but you can also use the SMuFL name with `glyph.name`
```xml
<tempo xml:id="tempo-L1F1" staff="1" tstamp="1.000000">
   <symbol glyph.name="metNoteQuarterUp" glyph.auth="smufl"/> = 152</tempo>
```

For question (2) I fixed it and it should not be crashing any more. You should now see a warning saying that the CSS font cannot be found. I suspect you are pointing to a resource directory that is not up-do-date with the Verovio version. Make sure it is.

**Comment by craigsapp** [comment]
The new suppressing of segmentation faults related to unknown fonts is working:

Old text encoding:
```
[Warning] Using rend@fontname with 'VerovioText' is deprecated. Use 'rend@fontfam="smufl"' instead
[Warning] The CSS font for 'Leipzig' could not be loaded and will not be embedded in the SVG
Output written to xold.svg.
```

New text encoding:
```
[Warning] The CSS font for 'Leipzig' could not be loaded and will not be embedded in the SVG
Output written to x.svg.
```

This is on the command-line.  Does the Leipzig font need to be installed for the OS (from verovio/fonts/Leipzig/Leipzig.ttf), or is it supposed to be from the `/usr/local/share/verovio` directory?  I have done`make install` (with the latest `develop` branch) and there is updated XML content for the Leipzig font in that directory.

After installing Leipzig.ttf as an OS font, there is still a complaint about not finding the CSS font `Leipzig`, but it will be displayed in the SVG image in a browser (since Leipzig is probably found in the MacOS font library):


<img width="273" alt="Screen Shot 2022-10-05 at 12 15 54 PM" src="https://user-images.githubusercontent.com/3487289/194143748-2eac2922-70ab-4f36-81f1-23350770f964.png">

Without installing Leipzig in the OS, the Leipzig character is a box:

<img width="242" alt="Screen Shot 2022-10-05 at 12 26 47 PM" src="https://user-images.githubusercontent.com/3487289/194145644-b354c07c-2233-4ae8-a1ad-c84ea82c6828.png">

Is there a new way to install fonts other than `make install`?  Or otherwise a new option to embed a font within the SVG (preferably only the used glyphs)?

**Reply by ahankinson** [reply]
If you've done `make install` from the latest branch, it will put it in the default cmake directory. (`CMAKE_PREFIX_DIR`, if I remember correctly). If you have homebrew installed, and cmake installed with homebrew, I think this means that it will install the fonts within `/opt/homebrew/share/verovio`. 

Otherwise, if none of these are set then it will default to `/usr/local/share/verovio`. 

The new font system base64-encodes the WOFF and puts it in a CSS file, which will be loaded (see: https://github.com/rism-digital/verovio/blob/develop/data/Leipzig.css)

I think, actually, that there's a bug in this, though -- it looks like it's only copying XML files, but I think it now also needs to copy the CSS. I'll see if I can put in a fix.

**Reply by ahankinson** [reply]
On line 250 of the CMakeLists.txt file, try adding `PATTERN "*.css"` to the end of the line, then re-run cmake, make, make install and see if that fixes it.

**Reply by ahankinson** [reply]
See #3056 for what I think is a fix. I tried rendering your sample file after applying this fix and it seemed to work fine for me.

I don't have any of the fonts installed as system fonts.

**Comment by craigsapp** [comment]
I notice that `<symbol>` cannot be placed inside of a `<rend>`, which might be nice thing to be able to do, so at the moment I will use:

```xml
<rend fontfam="smufl">&#xECA5;</rend>
```

Instead of 

```xml
<symbol glyph.name="metNoteQuarterUp" glyph.auth="smufl"/>
```

**Comment by craigsapp** [comment, marked as answer]
> See https://github.com/rism-digital/verovio/pull/3056 for what I think is a fix. I tried rendering your sample file after applying this fix and it seemed to work fine for me.

Thanks!  That fix worked.   After using it, the css fonts are installed in `/usr/local/share/verovio/*.css` and the SVG image has the font embedded at the end of the file:

<img width="650" alt="Screen Shot 2022-10-05 at 3 17 39 PM" src="https://user-images.githubusercontent.com/3487289/194174368-0df20847-43f3-4097-a752-b2e02203db7f.png">

Now the SVG image renders properly when the Leipzig font is not installed in the OS:

<img width="233" alt="Screen Shot 2022-10-05 at 3 19 39 PM" src="https://user-images.githubusercontent.com/3487289/194174435-b91de899-9b4f-4541-9d49-bda5ff17fe1c.png">

And of course the Warning about the missing font is no longer printed:

```
[Warning] The CSS font for 'Leipzig' could not be loaded and will not be embedded in the SVG
```
